File: books/dsa/graphs/topo_sort.py
from collections import deque, defaultdict
from typing import List, Tuple


# Pattern: Kahn topological sort.
# Invariant: queue contains nodes with zero remaining prerequisites.
# Complexity: O(V + E) time, O(V + E) space.
# Interview line: if we cannot consume every course, a dependency cycle remains.
def can_finish(num_courses: int, prerequisites: List[List[int]]) -> bool:
    graph = defaultdict(list)
    indegree = [0] * num_courses

    for course, prereq in prerequisites:
        graph[prereq].append(course)
        indegree[course] += 1

    q = deque(i for i, degree in enumerate(indegree) if degree == 0)
    taken = 0

    while q:
        node = q.popleft()
        taken += 1
        for nei in graph[node]:
            indegree[nei] -= 1
            if indegree[nei] == 0:
                q.append(nei)

    return taken == num_courses


# Pattern: topological ordering with indegrees.
# Invariant: append a node only after all prerequisites have been consumed.
# Complexity: O(V + E) time, O(V + E) space.
# Interview line: the order exists only if the topo process visits every node.
def find_course_order(num_courses: int, prerequisites: List[List[int]]) -> List[int]:
    graph = defaultdict(list)
    indegree = [0] * num_courses

    for course, prereq in prerequisites:
        graph[prereq].append(course)
        indegree[course] += 1

    q = deque(i for i, degree in enumerate(indegree) if degree == 0)
    order = []

    while q:
        node = q.popleft()
        order.append(node)
        for nei in graph[node]:
            indegree[nei] -= 1
            if indegree[nei] == 0:
                q.append(nei)

    return order if len(order) == num_courses else []


# Pattern: DFS color marking for directed cycle detection.
# Invariant: color 1 means currently on recursion stack, color 2 means fully processed.
# Complexity: O(V + E) time, O(V + E) space.
# Interview line: seeing a gray node again means we found a back edge and therefore a cycle.
def has_cycle_dfs(num_nodes: int, edges: List[Tuple[int, int]]) -> bool:
    graph = defaultdict(list)
    for a, b in edges:
        graph[a].append(b)

    color = [0] * num_nodes

    # Pattern: recursion-stack cycle helper.
    # Invariant: gray nodes are active ancestors in the current DFS path.
    # Complexity: O(outdegree(node)) local work, O(V) recursion stack overall.
    # Interview line: a gray neighbor means this path loops back to itself.
    def dfs(node: int) -> bool:
        color[node] = 1
        for nei in graph[node]:
            if color[nei] == 1:
                return True
            if color[nei] == 0 and dfs(nei):
                return True
        color[node] = 2
        return False

    return any(color[node] == 0 and dfs(node) for node in range(num_nodes))

File: books/dsa/graphs/topo_sort.go
package graphs

// Pattern: Kahn topological sort.
// Invariant: queue contains nodes with zero remaining prerequisites.
// Complexity: O(V + E) time, O(V + E) space.
// Interview line: if we cannot consume every course, a dependency cycle remains.
func CanFinish(numCourses int, prerequisites [][]int) bool {
	adj := make(map[int][]int)
	indegree := make([]int, numCourses)

	for _, pre := range prerequisites {
		course, prereq := pre[0], pre[1]
		adj[prereq] = append(adj[prereq], course)
		indegree[course]++
	}

	var q []int
	for i := 0; i < numCourses; i++ {
		if indegree[i] == 0 {
			q = append(q, i)
		}
	}

	visited := 0
	for len(q) > 0 {
		curr := q[0]
		q = q[1:]
		visited++

		for _, neighbor := range adj[curr] {
			indegree[neighbor]--
			if indegree[neighbor] == 0 {
				q = append(q, neighbor)
			}
		}
	}

	return visited == numCourses
}

// Pattern: topological ordering with indegrees.
// Invariant: append a node only after all prerequisites have been consumed.
// Complexity: O(V + E) time, O(V + E) space.
// Interview line: the order exists only if the topo process visits every node.
func FindOrder(numCourses int, prerequisites [][]int) []int {
	adj := make(map[int][]int)
	indegree := make([]int, numCourses)

	for _, pre := range prerequisites {
		course, prereq := pre[0], pre[1]
		adj[prereq] = append(adj[prereq], course)
		indegree[course]++
	}

	var q []int
	for i := 0; i < numCourses; i++ {
		if indegree[i] == 0 {
			q = append(q, i)
		}
	}

	var order []int
	for len(q) > 0 {
		curr := q[0]
		q = q[1:]
		order = append(order, curr)

		for _, neighbor := range adj[curr] {
			indegree[neighbor]--
			if indegree[neighbor] == 0 {
				q = append(q, neighbor)
			}
		}
	}

	if len(order) == numCourses {
		return order
	}
	return nil
}

// Pattern: DFS color marking for directed cycle detection.
// Invariant: color 1 means currently on recursion stack, color 2 means fully processed.
// Complexity: O(V + E) time, O(V + E) space.
// Interview line: seeing a gray node again means we found a back edge and therefore a cycle.
func HasCycleDFS(numCourses int, prerequisites [][]int) bool {
	adj := make(map[int][]int)
	for _, pre := range prerequisites {
		adj[pre[1]] = append(adj[pre[1]], pre[0])
	}

	// 0: unvisited, 1: visiting (gray), 2: visited (black)
	colors := make([]int, numCourses)

	var dfs func(node int) bool
	dfs = func(node int) bool {
		colors[node] = 1 // Mark visiting
		for _, neighbor := range adj[node] {
			if colors[neighbor] == 1 {
				return true // Cycle detected
			}
			if colors[neighbor] == 0 {
				if dfs(neighbor) {
					return true
				}
			}
		}
		colors[node] = 2 // Mark visited
		return false
	}

	for i := 0; i < numCourses; i++ {
		if colors[i] == 0 {
			if dfs(i) {
				return true
			}
		}
	}
	return false
}
